File: Core/Source/NSCoder+DTCompatibility.m
//
//  NSCoder+DTCompatibility.m
//  DTCoreText
//
//  Created by Ryan Johnson on 14/02/19.
//  Copyright (c) 2014 Drobnik.com. All rights reserved.
//

#import "NSCoder+DTCompatibility.h"

#if !TARGET_OS_IPHONE
NSString* NSStringFromNSEdgeInsets(NSEdgeInsets insets);
NSEdgeInsets NSEdgeInsetsFromString(NSString *string);
#endif

@implementation NSCoder (DTCompatibility)

#if !TARGET_OS_IPHONE
- (void)encodeCGSize:(CGSize)size forKey:(NSString *)key {
    [self encodeObject:NSStringFromCGSize(size) forKey:key];
}

- (CGSize)decodeCGSizeForKey:(NSString *)key {
	return NSSizeToCGSize(NSSizeFromString([self decodeObjectForKey:key]));
}
#endif

- (void)encodeDTEdgeInsets:(DTEdgeInsets)insets forKey:(NSString *)key {
#if TARGET_OS_IPHONE
	[self encodeUIEdgeInsets:insets forKey:key];
#else
	[self encodeObject:NSStringFromNSEdgeInsets(insets) forKey:key];
#endif
}

- (DTEdgeInsets)decodeDTEdgeInsetsForKey:(NSString *)key {
#if TARGET_OS_IPHONE
	return [self decodeUIEdgeInsetsForKey:key];
#else
	return NSEdgeInsetsFromString([self decodeObjectForKey:key]);
#endif
}


@end

#if !TARGET_OS_IPHONE
NSString* NSStringFromNSEdgeInsets(NSEdgeInsets insets) {
	return [NSString stringWithFormat:@"{%f,%f,%f,%f}", insets.top, insets.left, insets.bottom, insets.right];
}

NSEdgeInsets NSEdgeInsetsFromString(NSString *string) {
	// Cut off curly brackets
	string = [string substringWithRange:NSMakeRange(1, string.length - 2)];

	NSArray *floatStrings = [string componentsSeparatedByString:@","];
	if (floatStrings.count != 4) {
		return NSEdgeInsetsMake(0, 0, 0, 0);
	}
	CGFloat top = [floatStrings[0] floatValue];
	CGFloat left = [floatStrings[1] floatValue];
	CGFloat bottom = [floatStrings[2] floatValue];
	CGFloat right = [floatStrings[3] floatValue];

	return NSEdgeInsetsMake(top, left, bottom, right);
}

#endif

File: Sources/DTCoreTextSwift/NSCoder+DTCompatibility.swift
import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

extension NSCoder {

  #if canImport(AppKit) && !canImport(UIKit)
    /// Encodes a CGSize for the given key (macOS only; iOS has this built in).
    @objc public func encodeCGSize(_ size: CGSize, forKey key: String) {
      encode(NSStringFromSize(NSSizeFromCGSize(size)), forKey: key)
    }

    /// Decodes a CGSize for the given key (macOS only; iOS has this built in).
    @objc public func decodeCGSize(forKey key: String) -> CGSize {
      let string = decodeObject(forKey: key) as? String ?? ""
      return NSSizeToCGSize(NSSizeFromString(string))
    }
  #endif

  /// Encodes DTEdgeInsets for the given key.
  @objc public func encodeDTEdgeInsets(_ insets: DTEdgeInsets, forKey key: String) {
    #if canImport(UIKit)
      encode(insets, forKey: key)
    #else
      let string = String(
        format: "{%f,%f,%f,%f}", insets.top, insets.left, insets.bottom, insets.right)
      encode(string, forKey: key)
    #endif
  }

  /// Decodes DTEdgeInsets for the given key.
  @objc public func decodeDTEdgeInsets(forKey key: String) -> DTEdgeInsets {
    #if canImport(UIKit)
      return decodeUIEdgeInsets(forKey: key)
    #else
      guard let string = decodeObject(forKey: key) as? String else {
        return NSEdgeInsets(top: 0, left: 0, bottom: 0, right: 0)
      }
      return nsEdgeInsetsFromString(string)
    #endif
  }
}
